### PR TITLE
sygus python api

### PR DESCRIPTION
**Pull request by yoni206**
This PR extends the python api with support for sygus functionality. 
This required the addition of a nullary constructor for `Grammar` in the c++ api.
A unit test is also included, and is a translation of the corresponding c++ api unit test.

Examples are not added yet, but are ready and planned for a next PR (in order to keep this one shorter).

### CODE REVIEW

File: src/api/cvc4cpp.h [hunk 1]
```diff
@@ -1954,6 +1954,11 @@ class CVC4_PUBLIC Grammar
    */
   void addRules(Term ntSymbol, std::vector<Term> rules);
 
+  /**
+   * Nullary constructor. Needed for the python api.
```
Comment (makaimann) [inline src/api/cvc4cpp.h:1958]: Super minor, but lets call this the Cython API for clarity. Technically the old SWIG python bindings haven't been deprecated yet.

File: src/api/python/cvc4.pxi [hunk 2]
```diff
@@ -773,6 +793,63 @@ cdef class Solver:
...
+        cdef vector[c_Term] bvc
+        cdef vector[c_Term] ntc
+        for bv in boundVars:
+            bvc.push_back((<Term?> bv).cterm)
+        for nt in ntSymbols:
+            ntc.push_back((<Term?> nt).cterm)
+        grammar.cgrammar = self.csolver.mkSygusGrammar(<const vector[c_Term]&> bvc, <const vector[c_Term]&> ntc)
+        return grammar
+
+    def mkSygusVar(self, Sort sort, symbol=""):
```
Comment (makaimann) [inline src/api/python/cvc4.pxi:807]: Also minor, but you can type the symbol as a `str`, right?

File: src/api/python/cvc4.pxi [hunk 3]
```diff
@@ -773,6 +793,63 @@ cdef class Solver:
...
+        term.cterm = self.csolver.mkSygusVar(sort.csort, symbol.encode())
+        return term
+
+    def addSygusConstraint(self, Term t):
+        self.csolver.addSygusConstraint(t.cterm)
+
+    def addSygusInvConstraint(self, Term inv_f, Term pre_f, Term trans_f, Term post_f):
+        self.csolver.addSygusInvConstraint(inv_f.cterm, pre_f.cterm, trans_f.cterm, post_f.cterm)
+
+    def synthFun(self, symbol, bound_vars, Sort sort, Grammar grammar=None):
```
Comment (makaimann) [inline src/api/python/cvc4.pxi:818]: Could type the symbol as a `str` if you'd like. Also, I didn't realize that you could use `None` as a default argument even with a type declaration -- very cool!

File: src/api/cvc4cpp.h (hunk 1 above)
Comment (4tXJ7f) [reply src/api/cvc4cpp.h]: The old SWIG bindings are actually already removed on `master` :) I don't have a preference w.r.t. Cython vs Python API though.

Comment (makaimann) [reply src/api/cvc4cpp.h]: Oh great, didn't realize that! Either is fine then.

**Pull request merged**